File: personalizatio-sdk/src/main/java/com/personalizatio/stories/views/StoryView.java
package com.personalizatio.stories.views;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.util.Log;
import android.util.Pair;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ToggleButton;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.media3.common.C;
import androidx.media3.common.PlaybackException;
import androidx.media3.common.Player;
import androidx.media3.common.Tracks;
import androidx.media3.common.util.UnstableApi;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.personalizatio.R;
import com.personalizatio.SDK;
import com.personalizatio.stories.models.Slide;
import com.personalizatio.stories.models.Story;
import com.personalizatio.stories.views.storyItem.StoryItemView;

import java.util.HashMap;

@SuppressLint("ViewConstructor")
final class StoryView extends ConstraintLayout implements StoriesProgressView.StoriesListener, Player.Listener {

	private final StoriesView storiesView;
	private Story story;

	private long pressTime = 0L;
	private final static long LIMIT = 500L;
	private Runnable completeListener;
	private Runnable prevStoryListener;

	private StoriesProgressView storiesProgressView;
	private OnTouchListener onTouchListener;
	private ViewPager2 mViewPager;
	private boolean storiesStarted = false;
	private boolean prevFocusState = true;
	private boolean locked = false;
	private final HashMap<Integer, PagerHolder> holders = new HashMap<>();
	private ToggleButton mute;

	//Heading
	private final StoryDialog.OnProgressState stateListener;

	private Pair<Integer, Integer> viewPagerSize = null;

	@SuppressLint("ClickableViewAccessibility")
	public StoryView(StoriesView storiesView, StoryDialog.OnProgressState stateListener) {
		super(storiesView.getContext());
		this.storiesView = storiesView;
		this.stateListener = stateListener;

		inflate(getContext(), R.layout.story_view, this);
		setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));

		initViews();
		setupViews();
	}

	private void initViews() {
		storiesProgressView = findViewById(R.id.storiesProgressView);

		mViewPager = findViewById(R.id.storiesViewPager);

		mute = findViewById(R.id.mute);
	}

	@SuppressLint("ClickableViewAccessibility")
	private void setupViews() {
		onTouchListener = (v, event) -> {
			if( storiesStarted ) {
				switch( event.getAction() ) {
					case MotionEvent.ACTION_DOWN:
						pressTime = System.currentTimeMillis();
						pause();
//							setHeadingVisibility(GONE);
						return false;
					case MotionEvent.ACTION_UP:
						long now = System.currentTimeMillis();
						if( LIMIT < now - pressTime ) {
							resume();
						}
//							setHeadingVisibility(VISIBLE);
						return LIMIT < now - pressTime;
				}
			}
			return false;
		};

		storiesProgressView.setColor(Color.parseColor(storiesView.getSettings().background_progress));
		storiesProgressView.setStoriesListener(this);

		mViewPager.setClipToPadding(false);
		mViewPager.setClipChildren(false);
		mViewPager.setOffscreenPageLimit(1);
		mViewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
			@Override
			public void onPageSelected(int position) {
				var player = storiesView.getPlayer().getPlayer();
				if( player.isPlaying() || player.isLoading() ) {
					player.pause();
				}
				//При изменении слайда останавливаем видео на скрытых
				for( int i = 0; i < story.getSlidesCount(); i++ ) {
					PagerHolder holder = getHolder(i);
					if( holder != null ) {
						if( i != position ) {
							holder.release();
						}
					}
				}
				if( storiesStarted ) {
					SDK.track_story("view", storiesView.getCode(), story.getId(), story.getSlide(position).getId());
					playVideo();
				}
			}
		});

		storiesView.getPlayer().getPlayer().setVolume(storiesView.isMute() ? 0f : 1f);
		storiesView.setMuteListener(() -> {
			storiesView.getPlayer().getPlayer().setVolume(1f);
		});

		//Управление звуком
		mute.setOnClickListener(v -> {
			storiesView.muteVideo(mute.isChecked());
			storiesView.getPlayer().getPlayer().setVolume(mute.isChecked() ? 0f : 1f);
		});
		mute.setChecked(storiesView.isMute());
	}

	public void setStory(Story story, Runnable completeListener, Runnable prevStoryListener) {
		this.story = story;
		this.completeListener = completeListener;
		this.prevStoryListener = prevStoryListener;

		var slidesCount = story.getSlidesCount();
		storiesProgressView.setStoriesCount(slidesCount);
		mViewPager.setAdapter(new ViewPagerAdapter());
		//Хак, чтобы не срабатывал onPageSelected при открытии первой кампании
		mViewPager.setCurrentItem(story.getStartPosition() == 0 ? slidesCount : 0, false);
		//Устанавливаем позицию
		mViewPager.setCurrentItem(story.getStartPosition(), false);
	}

	@UnstableApi
	private void playVideo() {
		if( storiesStarted ) {
			PagerHolder holder = getHolder(mViewPager.getCurrentItem());
			if( holder != null ) {
				Slide slide = story.getSlide(mViewPager.getCurrentItem());
				if( slide.getType().equals("video") ) {
					slide.setPrepared(false);
					//Подготавливаем плеер
					var player = storiesView.getPlayer().getPlayer();
					player.addListener(this);
					holder.getStoryItem().getVideo().setVisibility(GONE);
					holder.getStoryItem().getImage().setAlpha(1f);
					holder.getStoryItem().getVideo().setPlayer(player);
					storiesView.getPlayer().prepare(slide.getBackground());
					storiesProgressView.pause();
					mute.setChecked(storiesView.isMute());
				}
			}
		}
	}

	@Override
	public void onPlaybackStateChanged(@Player.State int playbackState) {
		switch( playbackState ) {
			case Player.STATE_BUFFERING:
				storiesProgressView.pause();
				break;
			case Player.STATE_READY:
				storiesView.getPlayer().getPlayer().play();
				break;
		}
	}

	@Override
	public void onIsPlayingChanged(boolean isPlaying) {
		if( isPlaying ) {
			PagerHolder holder = getHolder(mViewPager.getCurrentItem());
			holder.getStoryItem().getVideo().setVisibility(VISIBLE);
			holder.getStoryItem().getImage().animate().alpha(0).setDuration(300);
			story.getSlide(mViewPager.getCurrentItem()).setPrepared(true);
			resume();
		}
	}

	@Override
	public void onTracksChanged(@NonNull Tracks tracks) {
		var contentDuration = storiesView.getPlayer().getPlayer().getContentDuration();

		if(contentDuration > 0) {
			int currentItem = mViewPager.getCurrentItem();
			story.getSlide(currentItem).setDuration(contentDuration);
			updateDuration(currentItem);
		}

		mute.setVisibility(tracks.isTypeSupported(C.TRACK_TYPE_AUDIO) ? VISIBLE : GONE);
	}

	@Override
	public void onPlayerError(PlaybackException error) {
		Log.e(SDK.TAG, "player error: " + error.getMessage() + ", story: " + story.getId());
		PagerHolder holder = getCurrentHolder();
		holder.getStoryItem().reload_layout.setVisibility(VISIBLE);

		holder.getStoryItem().reload.setOnClickListener((View) -> {
			Slide slide = story.getSlide(mViewPager.getCurrentItem());
			if( slide.getType().equals("video") ) {
				holder.getStoryItem().reload_layout.setVisibility(GONE);
				playVideo();
			} else {
				holder.getStoryItem().update(slide, mViewPager.getCurrentItem(), storiesView.getCode(), story.getId());
			}
		});

		//Добавляем авто-обновление
		holder.getStoryItem().reload.postDelayed(() -> holder.getStoryItem().reload.callOnClick(), 15000L);
	}

	@Override
	public void onVolumeChanged(float volume) {
		mute.setChecked(volume == 0);
	}

	class PagerHolder extends RecyclerView.ViewHolder {
		private StoryItemView storyItem;

		public PagerHolder(@NonNull View view) {
			super(view);

			storyItem = (StoryItemView) view;
			if(viewPagerSize == null) {
				var viewPagerHeight = mViewPager.getHeight();
				var params = (ConstraintLayout.LayoutParams) storiesProgressView.getLayoutParams();
				var viewPagerTopOffset = params.height + params.bottomMargin + params.topMargin;
				viewPagerSize = new Pair<>(viewPagerHeight, viewPagerTopOffset);
			}
			storyItem.setViewSize(viewPagerSize.first, viewPagerSize.second);
			storyItem.setOnPageListener(new StoryItemView.OnPageListener() {
				@Override
				public void onPrev() {
					previousSlide();
				}

				@Override
				public void onNext() {
					nextSlide();
				}

				@Override
				public void onPrepared(int position) {
					if( story.getSlide(position).getType().equals("image") && storiesStarted ) {
						try {
							storiesProgressView.resume();
						} catch(IndexOutOfBoundsException e) {
						}
					}
				}

				@Override
				public void onLocked(boolean lock) {
					locked = lock;
					stateListener.onState(!lock);
					if( lock ) {
						pause();
					} else {
						resume();
					}
				}
			});
		}

		public void bind(Slide slide, int position) {
			holders.put(position, this);
			mute.setVisibility(GONE);
			storyItem.update(slide, position, storiesView.getCode(), story.getId());
			storyItem.setOnTouchListener(onTouchListener);

			//Устанавливаем загрузку видео, если биндим текущий элемент
			if( position == mViewPager.getCurrentItem() ) {
				playVideo();
			}
		}

		public void release() {
			storyItem.release();
		}

		public StoryItemView getStoryItem() {
			return storyItem;
		}
	}

	class ViewPagerAdapter extends RecyclerView.Adapter<PagerHolder> {

		@NonNull
		@Override
		public PagerHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			return new PagerHolder(new StoryItemView(storiesView));
		}

		@Override
		public void onBindViewHolder(@NonNull PagerHolder holder, int position) {
			holder.bind(story.getSlide(position), position);
		}

		@Override
		public int getItemCount() {
			return story.getSlidesCount();
		}
	}

	@Override
	public void onWindowFocusChanged(boolean state) {
		super.onWindowFocusChanged(state);
		//Когда окно сворачивается, нужно остановить видео и прогресс
		if( storiesProgressView != null && storiesStarted && prevFocusState != state ) {
			prevFocusState = state;
			if( state ) {
				resume();
			} else {
				pause();
			}
		}
	}

	public void updateDurations() {
		var slidesCount = story.getSlidesCount();
		long[] durations = new long[slidesCount];
		for( int i = 0; i < slidesCount; i++ ) {
			durations[i] = story.getSlide(i).getDuration();
		}
		storiesProgressView.setStoriesCountWithDurations(durations);
	}

	private PagerHolder getCurrentHolder() {
		return getHolder(mViewPager.getCurrentItem());
	}

	private PagerHolder getHolder(int position) {
		return holders.get(position);
	}

	public void pause() {
		storiesProgressView.pause();
		storiesView.getPlayer().getPlayer().pause();
	}

	public void resume() {
		var slide = story.getSlide(mViewPager.getCurrentItem());

		if( !locked && slide.isPrepared() ) {
			storiesProgressView.resume();
			var player = storiesView.getPlayer().getPlayer();
			if(slide.getType().equals("video") && !player.isLoading() && !player.isPlaying() ) {
				player.play();
			}
		}
	}

	@Override
	public void onNext() {
		var startPosition = story.getStartPosition();
		if(startPosition + 1 >= story.getSlidesCount()) {
			onComplete();
			return;
		}
		startPosition++;
		story.setStartPosition(startPosition);
		mViewPager.setCurrentItem(startPosition, false);
		storiesProgressView.startStories(startPosition);
	}

	@Override
	public void onStart(int position) {
		//Для текущего слайда
		if( position == mViewPager.getCurrentItem() ) {
			Slide slide = story.getSlide(position);
			//Если видео еще подгружается, приостанавливаем таймер анимации
			if( !slide.isPrepared() ) {
				storiesProgressView.pause();
			}
		}
	}

	@Override
	public void onPrev() {
		var startPosition = story.getStartPosition();
		if( startPosition <= 0 ) {
			prevStoryListener.run();
			return;
		}
		startPosition--;
		story.setStartPosition(startPosition);
		mViewPager.setCurrentItem(startPosition, false);
		storiesProgressView.startStories(startPosition);
	}

	@Override
	public void onComplete() {
		story.setViewed(true);
		story.setStartPosition(0);
		updateDurations();
		completeListener.run();
	}

	public void updateDuration(int position) {
		storiesProgressView.updateStoryDuration(position, story.getSlide(position).getDuration());
	}

	public void startStories() {
		var startPosition = story.getStartPosition();

		if( !storiesStarted ) {
			storiesStarted = true;
			playVideo();
			SDK.track_story("view", storiesView.getCode(), story.getId(), story.getSlide(startPosition).getId());
		}

		if( !locked ) {
			updateDurations();
			storiesProgressView.startStories(startPosition);
			mViewPager.setCurrentItem(startPosition, false);
		}
	}

	public void stopStories() {
		storiesStarted = false;
		//Если кампания сториса не активна на экране, удаляем листенеры
		storiesView.getPlayer().getPlayer().removeListener(this);
		pause();
		storiesProgressView.destroy();
		PagerHolder holder = getCurrentHolder();
		if( holder != null ) {
			holder.release();
		}
	}

	private void previousSlide() {
		if( storiesStarted && !locked ) {
			updateDurations();
			onPrev();
		}
	}

	public void nextSlide() {
		if( storiesStarted && !locked ) {
			onNext();
		}
	}

	public void release() {
		storiesView.getPlayer().getPlayer().removeListener(this);
		storiesView.getPlayer().getPlayer().stop();
		for( int i = 0; i < story.getSlidesCount(); i++ ) {
			PagerHolder holder = getHolder(i);
			if( holder != null ) {
				holder.release();
			}
		}
	}
}

File: personalizatio-sdk/src/main/kotlin/com/personalizatio/stories/views/StoryView.kt
package com.personalizatio.stories.views

import android.annotation.SuppressLint
import android.graphics.Color
import android.util.Log
import android.util.Pair
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.ToggleButton
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.media3.common.C
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.Tracks
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.personalizatio.R
import com.personalizatio.SDK
import com.personalizatio.stories.StoryState
import com.personalizatio.stories.models.Slide
import com.personalizatio.stories.models.Story
import com.personalizatio.stories.views.StoriesProgressView.StoriesListener
import com.personalizatio.stories.views.storyItem.StoryItemView
import com.personalizatio.stories.views.storyItem.StoryItemView.OnPageListener

@SuppressLint("ViewConstructor")
internal class StoryView @SuppressLint("ClickableViewAccessibility") constructor(
    private val storiesView: StoriesView, //Heading
    private val storyStateListener: StoryDialog.OnStoryStateListener
) : ConstraintLayout(storiesView.context), StoriesListener, Player.Listener {
    private var story: Story? = null

    private var pressTime = 0L
    private var completeListener: Runnable? = null
    private var prevStoryListener: Runnable? = null

    private lateinit var storiesProgressView: StoriesProgressView
    private var onTouchListener: OnTouchListener? = null
    private lateinit var mViewPager: ViewPager2
    private var storiesStarted = false
    private var prevFocusState = true
    private var locked = false
    private val holders = HashMap<Int, PagerHolder>()
    private lateinit var mute: ToggleButton

    private var viewPagerSize: Pair<Int, Int>? = null

    init {
        inflate(context, R.layout.story_view, this)
        layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)

        initViews()
        setupViews();
    }

    private fun initViews() {
        storiesProgressView = findViewById(R.id.storiesProgressView)

        mViewPager = findViewById(R.id.storiesViewPager)

        mute = findViewById(R.id.mute)
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupViews() {
        onTouchListener = OnTouchListener { _: View?, event: MotionEvent ->
            if (storiesStarted) {
                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        pressTime = System.currentTimeMillis()
                        pause()
                        //							setHeadingVisibility(GONE);
                        return@OnTouchListener false
                    }

                    MotionEvent.ACTION_UP -> {
                        val now = System.currentTimeMillis()
                        if (LIMIT < now - pressTime) {
                            resume()
                        }
                        //							setHeadingVisibility(VISIBLE);
                        return@OnTouchListener LIMIT < now - pressTime
                    }
                }
            }
            false
        }

        storiesProgressView.color = Color.parseColor(storiesView.settings.background_progress)
        storiesProgressView.storiesListener = this

        mViewPager.clipToPadding = false
        mViewPager.clipChildren = false
        mViewPager.offscreenPageLimit = 1
        mViewPager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                val player = com.personalizatio.stories.Player.player
                player?.let {
                    if (player.isPlaying || player.isLoading) {
                        player.pause()
                    }
                }

                story?.let { story ->
                    for (i in 0 until story.slidesCount) {
                        if (i != position) {
                            getHolder(i)?.release()
                        }
                    }

                    if (storiesStarted) {
                        SDK.instance.trackStory(
                            event = "view",
                            code = storiesView.code,
                            storyId = story.id,
                            slideId = story.getSlide(position).id
                        )
                        playVideo()
                    }
                }
            }
        })

        com.personalizatio.stories.Player?.let { player ->
            player.player?.let { innerPlayer ->
                innerPlayer.volume = if (storiesView.isMute) 0f else 1f
                storiesView.muteListener = Runnable {
                    innerPlayer.volume = 1f
                }

                mute.setOnClickListener {
                    storiesView.muteVideo(mute.isChecked)
                    innerPlayer.volume = if (mute.isChecked) 0f else 1f
                }
            }
        }

        mute.isChecked = storiesView.isMute
    }

    fun setStory(
        story: Story,
        completeListener: Runnable,
        prevStoryListener: Runnable
    ) {
        this.story = story
        this.completeListener = completeListener
        this.prevStoryListener = prevStoryListener

        val slidesCount = story.slidesCount
        storiesProgressView.setStoriesCount(slidesCount)
        mViewPager.adapter = ViewPagerAdapter()

        // Hack to prevent onPageSelected from triggering when opening the first campaign
        mViewPager.setCurrentItem(if (story.startPosition == 0) slidesCount else 0, false)

        mViewPager.setCurrentItem(story.startPosition, false)
    }

    private fun playVideo() {
        if (storiesStarted) {
            val holder = getHolder(mViewPager.currentItem)

            if (holder != null) {
                val slide = story?.getSlide(mViewPager.currentItem)
                if (slide?.type == "video") {
                    slide.isPrepared = false

                    val player = com.personalizatio.stories.Player.player
                    player?.addListener(this)
                    holder.storyItem.video.visibility = GONE
                    holder.storyItem.image.alpha = 1f
                    holder.storyItem.video.player = player
                    storiesView.preparePlayer(slide.background)
                    storiesProgressView.pause()
                    mute.isChecked = storiesView.isMute
                }
            }
        }
    }

    override fun onPlaybackStateChanged(playbackState: @Player.State Int) {
        when (playbackState) {
            Player.STATE_BUFFERING -> storiesProgressView.pause()
            Player.STATE_READY -> com.personalizatio.stories.Player.player?.play()
        }
    }

    override fun onIsPlayingChanged(isPlaying: Boolean) {
        if (isPlaying) {
            val holder = getHolder(mViewPager.currentItem)
            holder?.let {
                holder.storyItem.video.visibility = VISIBLE
                holder.storyItem.image.animate().alpha(0f).setDuration(300)
            }
            story?.getSlide(mViewPager.currentItem)?.isPrepared = true
            resume()
        }
    }

    override fun onTracksChanged(tracks: Tracks) {
        val player = com.personalizatio.stories.Player.player
        var contentDuration = 0L
        if (player != null) {
            contentDuration = player.contentDuration
        }

        if (contentDuration > 0) {
            val currentItem = mViewPager.currentItem
            story?.getSlide(currentItem)?.duration = contentDuration
            updateDuration(currentItem)
        }

        mute.visibility = if (tracks.isTypeSupported(C.TRACK_TYPE_AUDIO)) VISIBLE else GONE
    }

    override fun onPlayerError(error: PlaybackException) {
        Log.e(SDK.TAG, "player error: " + error.message + ", story: " + story?.id)
        val holder = currentHolder
        holder?.storyItem?.let { storyItem ->
            storyItem.reloadLayout.visibility = VISIBLE
            storyItem.reload.setOnClickListener {
                val slide = story?.getSlide(mViewPager.currentItem)
                slide?.let {
                    if (slide.type == "video") {
                        holder.storyItem.reloadLayout.visibility = GONE
                        playVideo()
                    } else {
                        storiesView.code?.let { code ->
                            holder.storyItem.update(slide, mViewPager.currentItem, code, story!!.id)
                        }
                    }
                }
            }

            storyItem.reload.postDelayed({ holder.storyItem.reload.callOnClick() }, 15000L)
        }
    }

    override fun onVolumeChanged(volume: Float) {
        mute.isChecked = volume == 0f
    }

    internal inner class PagerHolder(view: View) : RecyclerView.ViewHolder(view) {
        val storyItem: StoryItemView = view as StoryItemView

        init {
            if (viewPagerSize == null) {
                val viewPagerHeight = mViewPager.height
                val params = storiesProgressView.layoutParams as LayoutParams
                val viewPagerTopOffset = params.height + params.bottomMargin + params.topMargin
                viewPagerSize = Pair(viewPagerHeight, viewPagerTopOffset)
            }
            storyItem.setViewSize(viewPagerSize!!.first, viewPagerSize!!.second)
            storyItem.setOnPageListener(object : OnPageListener {
                override fun onPrev() {
                    previousSlide()
                }

                override fun onNext() {
                    nextSlide()
                }

                override fun onPrepared(position: Int) {
                    if (story?.getSlide(position)?.type == "image" && storiesStarted) {
                        try {
                            storiesProgressView.resume()
                        } catch (_: IndexOutOfBoundsException) {
                        }
                    }
                }

                override fun onLocked(lock: Boolean) {
                    locked = lock
                    storyStateListener.onStoryStateChanged(
                        storyState = if(lock) StoryState.PAUSE else StoryState.RUNNING
                    )
                    if (lock) {
                        pause()
                    } else {
                        resume()
                    }
                }
            })
        }

        @SuppressLint("ClickableViewAccessibility")
        fun bind(slide: Slide?, position: Int) {
            holders[position] = this
            mute.visibility = GONE
            slide?.let { storyItem.update(slide, position, storiesView.code, story!!.id) }
            onTouchListener?.let { listener -> storyItem.setOnTouchListener(listener) }

            if (position == mViewPager.currentItem) {
                playVideo()
            }
        }

        fun release() {
            storyItem.release()
        }
    }

    internal inner class ViewPagerAdapter : RecyclerView.Adapter<PagerHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PagerHolder {
            return PagerHolder(StoryItemView(
                context = storiesView.context,
                code = storiesView.code,
                settings = storiesView.settings,
                itemClickListener = storiesView.itemClickListener,
                storyStateListener = storyStateListener)
            )
        }

        override fun onBindViewHolder(holder: PagerHolder, position: Int) {
            holder.bind(story?.getSlide(position), position)
        }

        override fun getItemCount(): Int {
            return story?.slidesCount ?: 0
        }
    }

    override fun onWindowFocusChanged(state: Boolean) {
        super.onWindowFocusChanged(state)

        if (storiesStarted && prevFocusState != state) {
            prevFocusState = state
            if (state) {
                resume()
            } else {
                pause()
            }
        }
    }

    fun updateDurations() {
        story?.let { story ->
            val slidesCount = story.slidesCount
            val durations = LongArray(slidesCount)
            for (i in 0 until slidesCount) {
                durations[i] = story.getSlide(i).duration
            }
            storiesProgressView.setStoriesCountWithDurations(durations)
        }
    }

    private val currentHolder: PagerHolder?
        get() = getHolder(mViewPager.currentItem)

    private fun getHolder(position: Int): PagerHolder? {
        return holders[position]
    }

    fun pause() {
        storiesProgressView.pause()
        com.personalizatio.stories.Player.player?.pause()
    }

    fun resume() {
        val slide = story?.getSlide(mViewPager.currentItem)

        slide?.let {
            if (!locked && slide.isPrepared) {
                storiesProgressView.resume()
                val player = com.personalizatio.stories.Player.player
                player?.let {
                    if (slide.type == "video" && !player.isLoading && !player.isPlaying) {
                        player.play()
                    }
                }
            }
        }
    }

    override fun onNext() {
        story?.let { story ->
            var startPosition = story.startPosition
            if (startPosition + 1 >= story.slidesCount) {
                onComplete()
                return
            }
            startPosition++
            story.startPosition = startPosition
            mViewPager.setCurrentItem(startPosition, false)
            storiesProgressView.startStories(startPosition)
        }
    }

    override fun onStart(position: Int) {
        if (position == mViewPager.currentItem) {
            val slide = story?.getSlide(position)

            if (slide?.isPrepared != true) {
                storiesProgressView.pause()
            }
        }
    }

    override fun onPrev() {
        story?.let { story ->
            var startPosition = story.startPosition
            if (startPosition <= 0) {
                prevStoryListener!!.run()
                return
            }
            startPosition--
            story.startPosition = startPosition
            mViewPager.setCurrentItem(startPosition, false)
            storiesProgressView.startStories(startPosition)
        }
    }

    override fun onComplete() {
        story?.let { story ->
            story.isViewed = true
            story.startPosition = 0
        }
        updateDurations()
        completeListener?.run()
    }

    fun updateDuration(position: Int) {
        storiesProgressView.updateStoryDuration(position, story?.getSlide(position)?.duration ?: 0)
    }

    fun startStories() {
        story?.let { story ->
            val startPosition = story.startPosition

            if (!storiesStarted) {
                storiesStarted = true
                playVideo()
                SDK.instance.trackStory(
                    event = "view",
                    code = storiesView.code,
                    storyId = story.id,
                    slideId = story.getSlide(startPosition).id
                )
            }

            if (!locked) {
                updateDurations()
                storiesProgressView.startStories(startPosition)
                mViewPager.setCurrentItem(startPosition, false)
            }
        }
    }

    fun stopStories() {
        storiesStarted = false

        com.personalizatio.stories.Player.player?.removeListener(this)
        pause()
        storiesProgressView.destroy()
        val holder = currentHolder
        holder?.release()
    }

    private fun previousSlide() {
        if (storiesStarted && !locked) {
            updateDurations()
            onPrev()
        }
    }

    fun nextSlide() {
        if (storiesStarted && !locked) {
            onNext()
        }
    }

    fun release() {
        com.personalizatio.stories.Player.player?.removeListener(this)
        com.personalizatio.stories.Player.player?.stop()
        val slidesCount = story?.slidesCount ?: 0
        for (i in 0 until slidesCount) {
            val holder = getHolder(i)
            holder?.release()
        }
    }

    companion object {
        private const val LIMIT = 500L
    }
}
